File: core/src/main/java/com/yopox/ui/Letter.java
package com.yopox.ui;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Pixmap;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.yopox.IntPair;
import com.yopox.Util;

import java.util.HashMap;
import java.util.Objects;
import java.util.Vector;

class LetterKey {
    private final int x;
    private final int y;
    private final Color fg;
    private final Color bg;

    public LetterKey(int y, int x, Color fg, Color bg) {
        this.x = x;
        this.y = y;
        this.fg = fg;
        this.bg = bg;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        LetterKey letterKey = (LetterKey) o;
        return x == letterKey.x && y == letterKey.y && fg.equals(letterKey.fg) && bg.equals(letterKey.bg);
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y, fg, bg);
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public Color getFg() {
        return fg;
    }

    public Color getBg() {
        return bg;
    }
}

public class Letter {
    public static final String VALID_CHARS = "abcdefghijklmnopqrstuvwxyzàâçèéêëîôùû0123456789";
    public static final String GIVEN_CHARS = ":;,-/\"()&%' ";
    private static final HashMap<LetterKey, Texture> myTextures = new HashMap<>();

    private char myChar;
    private LetterKey myKey;

    private Color fg;
    private Color bg;

    public Letter(char character) {
        this(character, Colors.DEFAULT_FG, Colors.DEFAULT_BG);
    }

    public Letter(char character, Color fg, Color bg) {
        myChar = character;
        this.fg = fg;
        this.bg = bg;
        computeKey();
    }

    private void computeKey() {
        final IntPair indexes = Util.indexes.getOrDefault(myChar, IntPair.DEFAULT);
        myKey = new LetterKey(indexes.first, indexes.second, fg, bg);
    }

    public char getCharacter() {
        return myChar;
    }

    public void setCharacter(char character) {
        this.myChar = character;
        computeKey();
    }

    public Color getFg() {
        return fg;
    }

    public void setFg(Color fg) {
        this.fg = fg;
        computeKey();
    }

    public Color getBg() {
        return bg;
    }

    public void setBg(Color bg) {
        this.bg = bg;
        computeKey();
    }

    public Texture getTexture() {
        return myTextures.computeIfAbsent(myKey, letterKey -> {
            final TextureRegion region = Text.mrmoRegions[myKey.getY()][myKey.getX()];
            region.getTexture().getTextureData().prepare();
            final Pixmap original = region.getTexture().getTextureData().consumePixmap();

            final Pixmap pixmap = new Pixmap(Util.TILE.intValue(), Util.TILE.intValue(), Pixmap.Format.RGBA8888);
            for (int y = 0; y < pixmap.getHeight(); y++) {
                for (int x = 0; x < pixmap.getWidth(); x++) {
                    final Color color = original.getPixel(x + letterKey.getX() * Util.TILE.intValue(), y + letterKey.getY() * Util.TILE.intValue()) == -1
                            ? fg
                            : bg;
                    pixmap.setColor(color);
                    pixmap.drawPixel(x, y);
                }
            }

            return new Texture(pixmap);
        });
    }

    public static Vector<Letter> listOf(String text, Color fg, Color bg) {
        final Vector<Letter> vector = new Vector<>();
        for (int i = 0; i < text.length(); i++) {
            vector.add(new Letter(Character.toLowerCase(text.charAt(i)), fg, bg));
        }
        return vector;
    }

    public static String stringOf(Vector<Letter> vector) {
        StringBuilder result = new StringBuilder();
        for (Letter letter : vector) {
            result.append(letter.getCharacter());
        }
        return result.toString();
    }

    public static Float heightOf(Vector<Letter> vector) {
        if (vector.isEmpty()) return 0f;
        return (stringOf(vector).split("\n").length * 2 - 1) * Util.TILE;
    }
}

File: core/src/main/kotlin/com/yopox/ui/Letter.kt
package com.yopox.ui

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.Pixmap
import com.badlogic.gdx.graphics.Texture
import com.yopox.TILE

private val textures = HashMap<Pair<Char, Color>, Texture>()

data class Letter(val char: Char, var color: Color = Color.WHITE) {
    fun indexes(): Pair<Int, Int> = when (char) {
        in 'a'..'z' -> 28 to char.code - 'a'.code + 1
        in 'A'..'Z' -> 28 to char.code - 'A'.code + 1
        in '0'..'9' -> 27 to 16 + char.code - '0'.code
        '!' -> 27 to 1
        '"' -> 27 to 2
        '&' -> 27 to 6
        '\'' -> 27 to 7
        '(' -> 27 to 8
        ')' -> 27 to 9
        ',' -> 27 to 12
        '-' -> 27 to 13
        '.' -> 27 to 14
        '/' -> 27 to 15
        ':' -> 27 to 26
        ';' -> 27 to 27
        'à', 'À' -> 29 to 20
        'â', 'Â' -> 29 to 22
        'è', 'È' -> 29 to 28
        'é', 'É' -> 29 to 29
        'ê', 'Ê' -> 29 to 30
        'ë', 'Ë' -> 29 to 31
        'ô', 'Ô' -> 30 to 8
        'ù', 'Ù' -> 30 to 12
        '?' -> 31 to 27
        else -> 0 to 0
    }

    fun texture(): Texture {
        textures[char to color]?.let { return it }

        val i = indexes()
        val texture = mrmoRegions[i.first][i.second]
        texture.texture.textureData.prepare()
        val pixmap = texture.texture.textureData.consumePixmap()
        val newPixmap = Pixmap(8, 8, Pixmap.Format.RGBA8888)

        for (y in 0 until pixmap.height)
            for (x in 0 until pixmap.width)
                if (pixmap.getPixel(
                        x + i.second * TILE.toInt(),
                        y + i.first * TILE.toInt()
                    ) == -1) {
                    newPixmap.setColor(color)
                    newPixmap.drawRectangle(x, y, 1, 1)
                }

        Texture(newPixmap).let {
            textures[char to color] = it
            return it
        }
    }
}

fun List<Letter>.string(): String {
    return map { it.char.toString() }.reduce { s1, s2 -> s1 + s2 }
}
